external-crates/move/move-compiler/src/sui_mode/typing.rs

```diff
@@ -42,90 +51,424 @@ impl TypingVisitor for SuiTypeChecks {
+
 //**************************************************************************************************
 // Entry
 //**************************************************************************************************
 
-pub fn program(env: &mut CompilationEnv, info: &ProgramInfo, prog: &T::Program) {
-    let T::Program { modules, scripts } = prog;
-    for script in scripts.values() {
-        let config = env.package_config(script.package_name);
-        if config.flavor != Flavor::Sui {
-            continue;
-        }
+impl<'a> TypingVisitorContext for Context<'a> {
+    fn add_warning_filter_scope(&mut self, filter: WarningFilters) {
+        self.env.add_warning_filter_scope(filter)
+    }
 
-        // TODO point to PTB docs?
-        let msg = "'scripts' are not supported on Sui. \
-        Consider removing or refactoring into a 'module'";
-        env.add_diag(diag!(SCRIPT_DIAG, (script.loc, msg)))
+    fn pop_warning_filter_scope(&mut self) {
+        self.env.pop_warning_filter_scope()
     }
-    for (mident, mdef) in modules.key_cloned_iter() {
-        module(env, info, mident, mdef);
+
+    fn visit_script_custom(&mut self, _name: Symbol, script: &mut T::Script) -> bool {
+        let config = self.env.package_config(script.package_name);
+        if config.flavor == Flavor::Sui {
+            // TODO point to PTB docs?
+            let msg = "'scripts' are not supported on Sui. \
+                        Consider removing or refactoring into a 'module'";
+            self.env.add_diag(diag!(SCRIPT_DIAG, (script.loc, msg)));
+        }
+        // skip scripts
+        true
     }
-}
 
-fn module(
-    env: &mut CompilationEnv,
-    info: &ProgramInfo,
-    mident: ModuleIdent,
-    mdef: &T::ModuleDefinition,
-) {
-    let config = env.package_config(mdef.package_name);
-    if config.flavor != Flavor::Sui {
-        return;
+    fn visit_module_custom(&mut self, ident: ModuleIdent, mdef: &mut T::ModuleDefinition) -> bool {
+        let config = self.env.package_config(mdef.package_name);
+        if config.flavor != Flavor::Sui {
+            // skip if not sui
+            return true;
+        }
+
+        if !mdef.is_source_module {
+            // Skip non-source, dependency modules
+            return true;
+        }
+
+        self.set_module(ident);
+        self.in_test = mdef.attributes.iter().any(|(_, attr_, _)| {
+            matches!(
+                attr_,
+                AttributeName_::Known(KnownAttribute::Testing(TestingAttribute::TestOnly))
+            )
+        });
+        if let Some(sdef) = mdef.structs.get_(&self.upper_module()) {
+            let valid_fields = if let N::StructFields::Defined(fields) = &sdef.fields {
+                invalid_otw_field_loc(fields).is_none()
+            } else {
+                true
+            };
+            if valid_fields {
+                let name = mdef.structs.get_full_key_(&self.upper_module()).unwrap();
```

Comment:
Would it make sense to factor out this specific operation into a function or otherwise signpost that this is the way we find our OTW?

Reply:
Not sure I follow. What is it you want to break out into a separate function?